spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/UDTValue.scala

```diff
@@ -8,10 +8,21 @@ import com.datastax.spark.connector.types.NullableTypeConverter
 
 final case class UDTValue(columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef])
   extends ScalaGettableData {
+
```

Comment:
We don't need metaData_ here. How about doing it this way:
```suggestion
final case class UDTValue(columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef])
  extends ScalaGettableData {

  private var metaData_ : Option[CassandraRowMetadata] = None

  lazy val metaData : CassandraRowMetadata = metaData_ match {
    case Some(x) => x
    case None => CassandraRowMetadata.fromColumnNames(columnNames)
  }

  def this(metaData: CassandraRowMetadata, columnValues: IndexedSeq[AnyRef]) = {
    this(metaData.columnNames, columnValues)
    this.metaData_ = Some(metaData)
  }

  override def productArity: Int = columnValues.size
  override def productElement(i: Int) = columnValues(i)
}

object UDTValue {

  def fromJavaDriverUDTValue(value: DriverUDTValue): UDTValue = {
    val fields = value.getType.getFieldNames.map(f => toName(f)).toIndexedSeq
    val values = fields.map(GettableData.get(value, _))
    UDTValue(fields, values)
  }

  def fromMap(map: Map[String, Any]): UDTValue =
    new UDTValue(map.keys.toIndexedSeq, map.values.map(_.asInstanceOf[AnyRef]).toIndexedSeq)

  val TypeTag = implicitly[TypeTag[UDTValue]]
  val Symbol = typeOf[UDTValue].asInstanceOf[TypeRef].sym

  implicit object UDTValueConverter extends NullableTypeConverter[UDTValue] {
    def targetTypeTag = TypeTag
    def convertPF = {
      case x: UDTValue => x
    }
  }

  def apply(metaData: CassandraRowMetadata, columnValues: IndexedSeq[AnyRef]): UDTValue =
    new UDTValue(metaData, columnValues)
}
```

Reply:
@jtgrabowski , yes, I wanted initially to go that way, but this will change the auto-generated 'unapply' method for the case class.

earlier one could have used:
```
p match {
 case UDTValue(columnName,columnValues) => ...
}
```
and if we change the case class, then the method allowing match/case to work will be also auto changed.

It seems that in scala it is possible to implement our own 'unapply' for the case class, e.g. to make it identical to what we had before this pull req. If you find it's reasonable to take care about 'unapply', I'll play with the custom 'unapply' method. Please, let me know.

Reply:
Yes, let's add the corresponding `unapply` method.

Reply:
@jtgrabowski I'll push the changes soon. Anyway, it seems that it is not possible to achieve the full binary compatibility with this change. I've recalled the 'copy' method of the case classes. its signature will also change. :(

Reply:
that's ok, let's change it anyway.

Reply:
done